File: prismpath-telemetry-rs/src/spiral.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Crystal Warden Supply Chain Labs LLC
//! Tier 6 — decision-first Fermat-spiral spatial packing (progressive, integer-only).

use crate::quantizer::{self, FieldPartition};
use crate::wire;
use crate::zeckendorf;
use prismpath_rs::{Graph, V};
use serde_json::json;
use std::collections::HashMap;

pub const GOLDEN_ANGLE_U32: u32 = 0x9E3779B9;
pub const GOLDEN_ANGLE_DEG: f64 = 180.0 * (3.0 - 2.23606797749979);

const ALWAYS: &[&str] = &["always", "true", "else", "otherwise", "default", "_"];
const NEVER: &[&str] = &["false", "never"];

pub fn theta_u32(spiral_index: u32) -> u32 {
    spiral_index.wrapping_mul(GOLDEN_ANGLE_U32)
}

pub fn radius2(spiral_index: u32) -> u32 {
    spiral_index
}

pub fn spiral_xy(spiral_index: usize, scale: f64) -> (f64, f64) {
    let radius = scale * (spiral_index as f64).sqrt();
    let theta = (spiral_index as f64 * GOLDEN_ANGLE_DEG).to_radians();
    (radius * theta.cos(), radius * theta.sin())
}

pub fn mixed_radix_gray(radices: &[usize]) -> Vec<Vec<usize>> {
    let n_fields = radices.len();
    if n_fields == 0 {
        return vec![];
    }
    let mut digits = vec![0isize; n_fields];
    let mut directions = vec![1isize; n_fields];
    let mut total = 1;
    for &radix in radices {
        total *= radix;
    }
    let mut out = Vec::with_capacity(total);
    out.push(digits.iter().map(|&digit| digit as usize).collect());

    for _ in 1..total {
        let mut position = (n_fields - 1) as isize;
        while position >= 0 {
            let idx = position as usize;
            let next_digit = digits[idx] + directions[idx];
            if next_digit >= 0 && (next_digit as usize) < radices[idx] {
                digits[idx] = next_digit;
                break;
            }
            directions[idx] = -directions[idx];
            position -= 1;
        }
        out.push(digits.iter().map(|&digit| digit as usize).collect());
    }
    out
}

/// The expression under a condition, trimmed the way the kernel trims.
///
/// Rust's `str::trim` strips Unicode White_Space, which is neither the kernel's set nor Python's:
/// it keeps U+001C to U+001F and strips U+FEFF. A condition padded with one of those classifies as
/// a `when` predicate in `is_deterministic` (which trims with py_trim) and would fall through to
/// the raw-expression branch here, so the spiral layout would drop fields the kernel routes on.
fn expr_of(cond: &str) -> String {
    let text = prismpath_rs::py_trim(cond);
    if text.to_lowercase().starts_with("when ") {
        prismpath_rs::py_trim(&text[5..]).to_string()
    } else {
        text.to_string()
    }
}

fn node_fields(graph: &Graph, node: &str, parts: &HashMap<String, FieldPartition>) -> Vec<String> {
    let mut seen = Vec::new();
    if let Some(flow_node) = graph.nodes.get(node) {
        for (_target, cond) in &flow_node.edges {
            if !prismpath_rs::is_deterministic(cond) || prismpath_rs::is_semantic(cond) {
                continue;
            }
            let expr = expr_of(cond);
            let lower = expr.to_lowercase();
            if ALWAYS.contains(&lower.as_str()) || NEVER.contains(&lower.as_str()) {
                continue;
            }
            for atom in quantizer::parse_atoms(&expr) {
                if parts.contains_key(&atom.field) && !seen.contains(&atom.field) {
                    seen.push(atom.field);
                }
            }
        }
    }
    seen.sort();
    seen
}

#[derive(Debug, Clone)]
pub struct SpiralLayout {
    pub graph: Graph,
    pub node: String,
    pub parts: HashMap<String, FieldPartition>,
    pub fields: Vec<String>,
    pub radices: Vec<usize>,
    pub routes: Vec<Option<String>>,
    pub band_index: HashMap<Option<String>, usize>,
    pub band_base: Vec<usize>,
    pub band_width: Vec<usize>,
    pub cell_of: Vec<Vec<usize>>,
    pub n_of: HashMap<Vec<usize>, usize>,
    pub size: usize,
}

impl SpiralLayout {
    pub fn new(graph: &Graph, node: &str) -> Result<Self, String> {
        let parts = quantizer::build_partitions(graph);
        let fields = node_fields(graph, node, &parts);
        if fields.is_empty() {
            return Err(format!(
                "node {:?} routes on no decision-relevant fields — nothing to pack",
                node
            ));
        }
        let radices: Vec<usize> = fields.iter().map(|field| parts[field].n).collect();

        let route_order = route_order_map(graph, node);

        let mut buckets: HashMap<Option<String>, Vec<Vec<usize>>> = HashMap::new();
        let mut bucket_order: Vec<Option<String>> = Vec::new();

        for cell in mixed_radix_gray(&radices) {
            let route = route_of_cell(graph, node, &parts, &fields, &cell);
            if !bucket_order.contains(&route) {
                bucket_order.push(route.clone());
            }
            buckets.entry(route).or_default().push(cell);
        }

        let mut routes = bucket_order;
        routes.sort_by_key(|route| route_order.get(route).copied().unwrap_or(route_order.len()));

        let mut band_index = HashMap::new();
        for (band, route) in routes.iter().enumerate() {
            band_index.insert(route.clone(), band);
        }

        let mut band_base = Vec::new();
        let mut band_width = Vec::new();
        let mut cell_of = Vec::new();
        let mut n_of = HashMap::new();
        let mut base = 0;

        for route in &routes {
            let cells = &buckets[route];
            band_base.push(base);
            band_width.push(cells.len());
            for cell in cells {
                n_of.insert(cell.clone(), cell_of.len());
                cell_of.push(cell.clone());
            }
            base += cells.len();
        }

        Ok(SpiralLayout {
            graph: graph.clone(),
            node: node.to_string(),
            parts,
            fields,
            radices,
            routes,
            band_index,
            band_base,
            band_width,
            cell_of,
            n_of,
            size: base,
        })
    }

    /// Panicking variant, std indexing style (`slice[i]` vs `slice.get(i)`); see [`Self::try_cell`].
    /// Precondition: the reading carries every spiral field with an in-partition value. The spiral
    /// is the Tier 6 layout layer, not the codec path.
    pub fn cell(&self, reading: &HashMap<String, V>) -> Vec<usize> {
        self.try_cell(reading)
            .expect("spiral reading value outside its partition")
    }

    /// Fallible twin of [`Self::cell`]: a missing spiral field or an out-of-partition value is an
    /// `Err`, never a panic.
    pub fn try_cell(&self, reading: &HashMap<String, V>) -> Result<Vec<usize>, String> {
        self.fields
            .iter()
            .map(|field| {
                let value = reading
                    .get(field)
                    .ok_or_else(|| format!("reading missing spiral field {field:?}"))?;
                self.parts[field].symbol(value)
            })
            .collect()
    }

    /// Panicking variant; see [`Self::try_index`].
    pub fn index(&self, reading: &HashMap<String, V>) -> usize {
        self.try_index(reading).expect("spiral reading value outside its partition")
    }

    pub fn try_index(&self, reading: &HashMap<String, V>) -> Result<usize, String> {
        let cell = self.try_cell(reading)?;
        self.n_of
            .get(&cell)
            .copied()
            .ok_or_else(|| format!("cell {cell:?} is not in the spiral layout"))
    }

    /// Panicking variant; see [`Self::try_band_id`].
    pub fn band_id(&self, reading: &HashMap<String, V>) -> usize {
        self.try_band_id(reading).expect("spiral reading value outside its partition")
    }

    pub fn try_band_id(&self, reading: &HashMap<String, V>) -> Result<usize, String> {
        let cell = self.try_cell(reading)?;
        let route = route_of_cell(&self.graph, &self.node, &self.parts, &self.fields, &cell);
        self.band_index
            .get(&route)
            .copied()
            .ok_or_else(|| format!("route {route:?} is not in the band layout"))
    }

    /// Panicking variant; see [`Self::try_route_of`].
    pub fn route_of(&self, spiral_index: usize) -> Option<String> {
        self.try_route_of(spiral_index).unwrap_or_else(|error| panic!("{}", error))
    }

    /// Fallible twin of [`Self::route_of`]: an index outside the spiral is an `Err`.
    pub fn try_route_of(&self, spiral_index: usize) -> Result<Option<String>, String> {
        for band in 0..self.routes.len() {
            if spiral_index < self.band_base[band] + self.band_width[band] {
                return Ok(self.routes[band].clone());
            }
        }
        Err(format!("index {} outside the spiral ({} cells)", spiral_index, self.size))
    }

    pub fn band_bounds(&self) -> Vec<(usize, usize, Option<String>)> {
        (0..self.routes.len())
            .map(|band| {
                (
                    self.band_base[band],
                    self.band_base[band] + self.band_width[band],
                    self.routes[band].clone(),
                )
            })
            .collect()
    }

    /// Panicking variant; see [`Self::try_reconstruct_band`].
    pub fn reconstruct_band(&self, band_id: usize) -> HashMap<String, V> {
        self.try_reconstruct_band(band_id)
            .unwrap_or_else(|error| panic!("{}", error))
    }

    pub fn try_reconstruct_band(&self, band_id: usize) -> Result<HashMap<String, V>, String> {
        let base = *self
            .band_base
            .get(band_id)
            .ok_or_else(|| format!("band {} outside the layout ({} bands)", band_id, self.routes.len()))?;
        Ok(cell_reading(&self.parts, &self.fields, &self.cell_of[base]))
    }

    /// Panicking variant; see [`Self::try_reconstruct`].
    pub fn reconstruct(&self, spiral_index: usize) -> HashMap<String, V> {
        self.try_reconstruct(spiral_index).unwrap_or_else(|error| panic!("{}", error))
    }

    pub fn try_reconstruct(&self, spiral_index: usize) -> Result<HashMap<String, V>, String> {
        let cell = self
            .cell_of
            .get(spiral_index)
            .ok_or_else(|| format!("index {} outside the spiral ({} cells)", spiral_index, self.size))?;
        Ok(cell_reading(&self.parts, &self.fields, cell))
    }

    /// Panicking variant; see [`Self::try_encode_decision`].
    pub fn encode_decision(&self, reading: &HashMap<String, V>) -> String {
        self.try_encode_decision(reading)
            .expect("spiral reading value outside its partition")
    }

    pub fn try_encode_decision(&self, reading: &HashMap<String, V>) -> Result<String, String> {
        // band_id + 1 >= 1, and zeckendorf::encode only errs for inputs < 1 — infallible here.
        zeckendorf::encode(self.try_band_id(reading)? + 1)
    }

    pub fn decode_decision(&self, bits: &str) -> Result<Option<String>, String> {
        // `bits` is untrusted wire data: a crafted code can decode to any index. checked_sub guards
        // the 1-based underflow and `.get` guards the out-of-range read — neither may panic here.
        let band = zeckendorf::decode(bits)?
            .checked_sub(1)
            .ok_or_else(|| "decoded band index 0 is invalid (codes are 1-based)".to_string())?;
        self.routes
            .get(band)
            .cloned()
            .ok_or_else(|| format!("decoded band index {band} is outside the layout ({} bands)", self.routes.len()))
    }

    /// Panicking variant; see [`Self::try_encode_progressive`].
    pub fn encode_progressive(&self, reading: &HashMap<String, V>) -> (String, String) {
        self.try_encode_progressive(reading)
            .expect("spiral reading value outside its partition")
    }

    pub fn try_encode_progressive(&self, reading: &HashMap<String, V>) -> Result<(String, String), String> {
        let spiral_index = self.try_index(reading)?;
        let route = self.try_route_of(spiral_index)?;
        let band = *self
            .band_index
            .get(&route)
            .ok_or_else(|| format!("route {route:?} is not in the band layout"))?;
        let local = spiral_index - self.band_base[band];
        // band + 1 and local + 1 are both >= 1; encode only errs for inputs < 1.
        Ok((zeckendorf::encode(band + 1)?, zeckendorf::encode(local + 1)?))
    }

    pub fn decode_progressive(
        &self,
        decision_bits: &str,
        refine_bits: &str,
    ) -> Result<HashMap<String, V>, String> {
        let band = zeckendorf::decode(decision_bits)? - 1;
        let local = zeckendorf::decode(refine_bits)? - 1;
        Ok(self.reconstruct(self.band_base[band] + local))
    }

    pub fn tessellation(&self) -> serde_json::Value {
        let cells: Vec<serde_json::Value> = self
            .cell_of
            .iter()
            .enumerate()
            .map(|(spiral_index, cell)| {
                let (x_coordinate, y_coordinate) = spiral_xy(spiral_index, 1.0);
                json!({
                    "cell": cell,
                    "n": spiral_index,
                    "band": self.band_index[&self.route_of(spiral_index)],
                    "route": self.route_of(spiral_index),
                    "xy": [(x_coordinate * 1e6).round() / 1e6, (y_coordinate * 1e6).round() / 1e6]
                })
            })
            .collect();

        let bands: Vec<serde_json::Value> = self
            .routes
            .iter()
            .enumerate()
            .map(|(band, route)| {
                json!({
                    "route": route,
                    "base": self.band_base[band],
                    "width": self.band_width[band]
                })
            })
            .collect();

        json!({
            "node": self.node,
            "fields": self.fields,
            "radices": self.radices,
            "bands": bands,
            "size": self.size,
            "cells": cells
        })
    }
}

fn route_order_map(graph: &Graph, node: &str) -> HashMap<Option<String>, usize> {
    let mut appear: Vec<Option<String>> = Vec::new();
    if let Some(flow_node) = graph.nodes.get(node) {
        for (target, cond) in &flow_node.edges {
            if prismpath_rs::is_deterministic(cond) {
                let t_opt = Some(target.clone());
                if !appear.contains(&t_opt) {
                    appear.push(t_opt);
                }
            }
        }
    }
    appear.reverse();
    let mut map = HashMap::new();
    for (order, target) in appear.into_iter().enumerate() {
        map.insert(target, order);
    }
    map
}

fn cell_reading(
    parts: &HashMap<String, FieldPartition>,
    fields: &[String],
    cell: &[usize],
) -> HashMap<String, V> {
    fields
        .iter()
        .zip(cell.iter())
        .map(|(field, &symbol)| (field.clone(), parts[field].representative(symbol)))
        .collect()
}

fn route_of_cell(
    graph: &Graph,
    node: &str,
    parts: &HashMap<String, FieldPartition>,
    fields: &[String],
    cell: &[usize],
) -> Option<String> {
    let reading = cell_reading(parts, fields, cell);
    wire::route_node(graph, node, &reading)
}

#[cfg(test)]
mod tests {
    use super::expr_of;

    #[test]
    fn expr_of_trims_the_whitespace_the_kernel_trims() {
        // U+001C is Python whitespace and so is stripped by is_deterministic's py_trim, but it is
        // not Unicode White_Space, so str::trim used to leave it and hide the predicate.
        let cond = "\u{1c}when temp > 50";
        assert!(prismpath_rs::is_deterministic(cond));
        assert_eq!(expr_of(cond), "temp > 50");
        assert_eq!(expr_of("  when temp > 50  "), "temp > 50");
        assert_eq!(expr_of("looks like a refund"), "looks like a refund");
    }
}

File: prismpath/telemetry/spiral.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 Crystal Warden Supply Chain Labs LLC
"""Tier 6  -  decision-first Fermat-spiral spatial packing (progressive, integer-only).

Packs a multi-variable reading into a **single ordered index** ``n`` whose contiguous ranges *are* the
routes, so the cheap wire quantity is a band ID ("transmit the decision, not the magnitude"), with a
within-band refinement recovered on demand. Two properties the plain per-field wire lacks:

  * **Decision-first layout.** The bands are the flow's own routes, laid out center-outward with the
    fallthrough/baseline (matched last in a first-match flow) at the dense center and the most-specific /
    severe branches (declared first) outward  -  the doc's "center = baseline, outward = deviation". Band
    membership is
    ``band_base[b] <= n < band_base[b] + width[b]``  -  two integer compares, a Level M atom  -  because a
    Fermat/Vogel spiral has ``r^2 = c^2 * n``, so a radial ring ``r < R`` *is* ``n < K``. The golden angle
    ``theta = n * 137.5deg`` is a deterministic function of ``n`` (a ``u32`` add, ``0x9E3779B9``, no trig),
    so the wire quantity stays a single integer: a 1-D ordered index with uniform 2-D coverage, not 2-D
    data.

  * **Progressive transmit order.** The band ID alone routes correctly at ~1 symbol (it is decision-
    lossless, magnitude-lossy); the within-band **Gray**-ordered local index refines it to the exact
    quantized cell (full magnitude) only when the link can afford it. On a degrading link you keep routing
    correctly and lose only fidelity.

All continuous math (``sqrt``, the golden angle in degrees) lives at **build time** here  -  the edge path
is integer table lookup + integer compares. This module builds the layout and the codec; the gate is the
routing-accuracy-vs-bits benchmark (``bench/spiral_bench.py``) and the decisions-preserved + frozen
tessellation conformance tests. Hysteresis at the band edges (anti-flap) is a sampling concern of the
edge, deliberately outside this codec.

Scope: an OPTION for *multi-dimensional, correlated* state where progressive refinement pays  -  not the
default. Scalar channels stay on the layer-2 quantizer (``quantizer.py`` / ``wire.py``); O(1) here holds
only after quantization to a bounded key space.
"""
from __future__ import annotations

import ast
import math
from typing import Any, Dict, Iterator, List, Optional, Tuple

from prismpath.kernel import predicates
from prismpath.telemetry import quantizer
from prismpath.telemetry import wire
from prismpath.telemetry import zeckendorf as zeck

GOLDEN_ANGLE_U32 = 0x9E3779B9                 # golden ratio * 2^32; add-with-overflow == mod 2*pi, no trig
GOLDEN_ANGLE_DEG = 180.0 * (3.0 - math.sqrt(5.0))   # ~= 137.5077640; build-time only (geometry/plots)


# --------------------------------------------------------------- integer-only spiral geometry (edge path)
def theta_u32(point_index: int) -> int:
    """Golden angle of point ``n`` as a ``u32`` phase  -  a single multiply-add, overflow == mod 2*pi."""
    return (point_index * GOLDEN_ANGLE_U32) & 0xFFFFFFFF


def radius2(point_index: int) -> int:
    """Squared radius of point ``n`` (``r^2 = c^2 * n``, ``c = 1``): a radial ring ``r < R`` is ``n < R^2``."""
    return point_index


def spiral_xy(point_index: int, scale: float = 1.0) -> Tuple[float, float]:
    """Build-time Cartesian coords of point ``n`` (Vogel model). Floats  -  for tessellation/plots only."""
    radius = scale * math.sqrt(point_index)
    theta = math.radians(point_index * GOLDEN_ANGLE_DEG)
    return (radius * math.cos(theta), radius * math.sin(theta))


# --------------------------------------------------------------- mixed-radix reflected Gray code
def mixed_radix_gray(radices: List[int]) -> Iterator[Tuple[int, ...]]:
    """Yield every cell of a mixed-radix grid in reflected Gray order: **consecutive tuples differ by 1
    in exactly one field**. Locality ordering for the on-demand within-band refinement (small deltas)."""
    field_count = len(radices)
    digits = [0] * field_count
    directions = [1] * field_count
    total = 1
    for radix in radices:
        total *= radix
    yield tuple(digits)
    for _ in range(total - 1):
        digit_index = field_count - 1
        while digit_index >= 0:
            nd = digits[digit_index] + directions[digit_index]
            if 0 <= nd < radices[digit_index]:
                digits[digit_index] = nd
                break
            # reflect this wheel, carry to the next more-significant one
            directions[digit_index] = -directions[digit_index]
            digit_index -= 1
        yield tuple(digits)


def _node_fields(graph, node: str, parts: Dict[str, "q.FieldPartition"]) -> List[str]:
    """The decision-relevant fields this node actually routes on (canonical sorted order)."""
    seen: List[str] = []
    for _target, cond in graph.nodes[node].edges:
        if not predicates.is_deterministic(cond) or predicates.is_semantic(cond):
            continue
        expr = predicates._expr_of(cond)
        if expr.lower() in predicates.ALWAYS or expr.lower() in predicates.NEVER:
            continue
        try:
            body = ast.parse(expr, mode="eval").body
        except SyntaxError:
            continue
        for field, _op, _const in quantizer.atoms_of(body):
            if field in parts and field not in seen:
                seen.append(field)
    return sorted(seen)


# --------------------------------------------------------------- the layout
class SpiralLayout:
    """A decision-first spiral packing of one node's joint quantized-cell space.

    Bands (routes) are contiguous, ordered center-outward (baseline central, severe branches outward);
    cells within a band are Gray-ordered for locality. ``index`` packs a reading to ``n``; ``band_id`` is
    the cheap decision symbol; ``route_of`` recovers the route from ``n`` with integer compares (the Level
    M atom).
    """

    def __init__(self, graph, node: str):
        self.graph = graph
        self.node = node
        self.parts = quantizer.build_partitions(graph)
        self.fields = _node_fields(graph, node, self.parts)
        if not self.fields:
            raise ValueError(f"node {node!r} routes on no decision-relevant fields  -  nothing to pack")
        self.radices = [self.parts[field].n for field in self.fields]

        # Route every joint cell; group cells by route in edge-declaration (severity) order, Gray-ordered
        # within each band -> contiguous, severity-ordered index ranges.
        order = self._route_order()                       # route -> severity rank (first appearance)
        buckets: Dict[Optional[str], List[Tuple[int, ...]]] = {}
        for cell in mixed_radix_gray(self.radices):       # Gray order => within-band order is Gray-local
            route = self._route_of_cell(cell)
            buckets.setdefault(route, []).append(cell)

        self.routes: List[Optional[str]] = sorted(buckets, key=lambda route: order.get(route, len(order)))
        self.band_index: Dict[Optional[str], int] = {route: band_index
                                                     for band_index, route in enumerate(self.routes)}
        self.band_base: List[int] = []
        self.band_width: List[int] = []
        self.cell_of: List[Tuple[int, ...]] = []          # n -> cell
        self.n_of: Dict[Tuple[int, ...], int] = {}        # cell -> n
        base = 0
        for route in self.routes:
            cells = buckets[route]
            self.band_base.append(base)
            self.band_width.append(len(cells))
            for cell in cells:
                self.n_of[cell] = len(self.cell_of)
                self.cell_of.append(cell)
            base += len(cells)
        self.size = base

    # -- build-time helpers -------------------------------------------------
    def _route_order(self) -> Dict[Optional[str], int]:
        """Severity rank per route: LOWER = more central. First-match flows declare severe branches first
        and the baseline (``else``) last, so we reverse edge order -> baseline at the dense center (rank 0),
        most-specific branches outward. An unrouted cell (``None``) sorts outermost."""
        appear: List[Optional[str]] = []
        for target, cond in self.graph.nodes[self.node].edges:
            if predicates.is_deterministic(cond) and target not in appear:
                appear.append(target)
        return {target: rank for rank, target in enumerate(reversed(appear))}

    def _cell_reading(self, cell: Tuple[int, ...]) -> Dict[str, Any]:
        return {field: self.parts[field].representative(symbol) for field, symbol in zip(self.fields, cell)}

    def _route_of_cell(self, cell: Tuple[int, ...]) -> Optional[str]:
        return wire.route_node(self.graph, self.node, self._cell_reading(cell))

    # -- packing (edge path) ------------------------------------------------
    def cell(self, reading: Dict[str, Any]) -> Tuple[int, ...]:
        """A reading -> its joint symbol tuple over this node's fields."""
        return tuple(self.parts[field].symbol(reading[field]) for field in self.fields)

    def index(self, reading: Dict[str, Any]) -> int:
        """A reading -> its spiral index ``n``."""
        return self.n_of[self.cell(reading)]

    def band_id(self, reading: Dict[str, Any]) -> int:
        """A reading -> its band (route) index  -  the cheap, decision-lossless wire symbol."""
        return self.band_index[self._route_of_cell(self.cell(reading))]

    def route_of(self, spiral_index: int) -> Optional[str]:
        """``n`` -> route, by integer band-boundary compares (``base <= n < base+width``: the Level M atom)."""
        if spiral_index < 0:
            # the bands are contiguous and ascending from 0, so the scan below would hand a negative
            # index the first band's route instead of refusing it
            raise ValueError(f"index {spiral_index} outside the spiral ({self.size} cells)")
        for band in range(len(self.routes)):
            if spiral_index < self.band_base[band] + self.band_width[band]:   # contiguous & ascending
                return self.routes[band]
        raise ValueError(f"index {spiral_index} outside the spiral ({self.size} cells)")

    def band_bounds(self) -> List[Tuple[int, int, Optional[str]]]:
        """``[(lo, hi_exclusive, route), ...]``  -  the contiguous decision-bands."""
        return [(self.band_base[band], self.band_base[band] + self.band_width[band], self.routes[band])
                for band in range(len(self.routes))]

    # -- reconstruct --------------------------------------------------------
    def reconstruct_band(self, band_id: int) -> Dict[str, Any]:
        """Band ID -> a representative reading of that band (routes to the band's route; magnitude-lossy)."""
        return self._cell_reading(self.cell_of[self.band_base[band_id]])

    def reconstruct(self, spiral_index: int) -> Dict[str, Any]:
        """``n`` -> the exact cell's representative reading (full quantized magnitude)."""
        return self._cell_reading(self.cell_of[spiral_index])

    # -- wire codec ---------------------------------------------------------
    def encode_decision(self, reading: Dict[str, Any]) -> str:
        """The cheap stream: just the band ID (Fibonacci-coded). Routes correctly; carries no magnitude."""
        return zeck.encode(self.band_id(reading) + 1)

    def decode_decision(self, bits: str) -> Optional[str]:
        """Cheap-stream bits -> the route."""
        return self.routes[zeck.decode(bits) - 1]

    def encode_progressive(self, reading: Dict[str, Any]) -> Tuple[str, str]:
        """(decision bits, refinement bits): band ID first, then the within-band Gray local index."""
        spiral_index = self.index(reading)
        band = self.band_index[self.route_of(spiral_index)]
        local = spiral_index - self.band_base[band]
        return zeck.encode(band + 1), zeck.encode(local + 1)

    def decode_progressive(self, decision_bits: str, refine_bits: str) -> Dict[str, Any]:
        """(decision, refinement) -> the exact cell's representative reading."""
        band = zeck.decode(decision_bits) - 1
        local = zeck.decode(refine_bits) - 1
        return self.reconstruct(self.band_base[band] + local)

    # -- conformance / inspection ------------------------------------------
    def tessellation(self) -> Dict[str, Any]:
        """A frozen-corpus view: every cell -> its ``n``, band index, route, and build-time xy."""
        cells = []
        for spiral_index, cell in enumerate(self.cell_of):
            x_coord, y_coord = spiral_xy(spiral_index)
            cells.append({"cell": list(cell), "n": spiral_index,
                          "band": self.band_index[self.route_of(spiral_index)],
                          "route": self.route_of(spiral_index),
                          "xy": [round(x_coord, 6), round(y_coord, 6)]})
        return {"node": self.node, "fields": self.fields, "radices": self.radices,
                "bands": [{"route": route, "base": self.band_base[band], "width": self.band_width[band]}
                          for band, route in enumerate(self.routes)],
                "size": self.size, "cells": cells}
